### PR TITLE
[rllib] Support torch device and distributions.

### LINKED ISSUE #4333: [rllib] PyTorch A2C is not GPU accelerated
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.6 LTS (GCP)
- **Ray installed from (source or binary)**: source
- **Ray version**: 0.6.4
- **Python version**: 3.6.8

### Describe the problem

I tested atari-a2c with tuned parameter(/tuned_examples/atari-a2c.yaml)
It showed great result for atari breakout.
However if "use_pytorch": true was added, the result is quite different. (I used "atari-a2c-pytorch.yaml" in the 'Source code / logs' section)
It was very slow and it seems that there was no improvement. 
I guess there's some performance issue on pytorch a2c, but are there any necessary options for pytorch a2c?

image
image

### Source code / logs

atari-a2c.yaml
```
atari-a2c:
    env:
        grid_search:
            - BreakoutNoFrameskip-v4
            - BeamRiderNoFrameskip-v4
            - QbertNoFrameskip-v4
            - SpaceInvadersNoFrameskip-v4
    run: A2C
    config:
        sample_batch_size: 20
        clip_rewards: True
        num_workers: 5
        num_envs_per_worker: 5
        num_gpus: 1
        lr_schedule: [
            [0, 0.0007],
            [20000000, 0.000000000001],
        ]
```
atari-a2c-pytorch.yaml
```
atari-a2c:
    env:
        grid_search:
            - BreakoutNoFrameskip-v4
            - BeamRiderNoFrameskip-v4
            - QbertNoFrameskip-v4
            - SpaceInvadersNoFrameskip-v4
    run: A2C
    config:
        sample_batch_size: 20
        clip_rewards: True
        num_workers: 5
        num_envs_per_worker: 5
        num_gpus: 1
        lr_schedule: [
            [0, 0.0007],
            [20000000, 0.000000000001],
        ]
        use_pytorch: True
```

### PR DESCRIPTION
**Pull request by cffan**
## What do these changes do?

- Move torch model and inputs to GPU if specified.
- Wrapper class for torch native distributions.
- Support "grad_clip" in A3C.
- Return grad_info in A3C.
- Fix A3C and PG log_prob calculated not using actions.

## Related issue number

Closes https://github.com/ray-project/ray/issues/4333

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by FlyClover** [comment]
The test environment is using pytorch-cpu. Should I change it to pytorch?

**Comment by ericl** [comment]
Hm, does that make a gpu available? AFAIK, none of our tests are currently run with GPUs. What is the limitation of pytorch-cpu?

**Comment by FlyClover** [comment]
Never mind. CPU case should be handled correctly now.

File: python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py [hunk 1]
```diff
@@ -17,25 +18,31 @@
...
-        action_log_probs = log_probs.gather(1, actions.view(-1, 1))
-        entropy = -(log_probs * probs).sum(-1).sum()
-        pi_err = -advantages.dot(action_log_probs.reshape(-1))
-        value_err = F.mse_loss(values.reshape(-1), value_targets)
+    def forward(self, policy_model, observations, actions, advantages,
+                value_targets):
+        logits, _, values, _ = policy_model(
+            {SampleBatch.CUR_OBS: observations}, [])
+        logits = logits
+        values = values
```
Comment (ericl) [inline python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]: These two lines seem redundant?

File: python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py [hunk 2]
```diff
@@ -17,25 +18,31 @@
...
+        dist = self.dist_class(logits)
+        log_probs = dist.logp(actions)
+        if len(log_probs.shape) > 1:
+            log_probs = log_probs.sum(-1)
+        self.entropy = dist.entropy().mean().cpu()
+        self.pi_err = -advantages.dot(log_probs.reshape(-1)).cpu()
+        self.value_err = F.mse_loss(values.reshape(-1), value_targets).cpu()
```
Comment (ericl) [inline python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]: Is it necessary to move the loss to cpu?

File: python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py [hunk 3]
```diff
@@ -66,29 +73,61 @@ class A3CTorchPolicyGraph(A3CPostprocessing, TorchPolicyGraph):
...
             loss,
             loss_inputs=[
                 SampleBatch.CUR_OBS, SampleBatch.ACTIONS,
                 Postprocessing.ADVANTAGES, Postprocessing.VALUE_TARGETS
-            ])
+            ],
+            action_distribution_cls=dist_class)
+
+    @override(PolicyGraph)
+    def compute_gradients(self, postprocessed_batch):
```
Comment (ericl) [inline python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]: Could we keep this method impl in TorchPolicyGraph?

File: python/ray/rllib/agents/pg/torch_pg_policy_graph.py [hunk 4]
```diff
@@ -17,15 +16,18 @@
...
-        log_probs = F.log_softmax(logits, dim=1)
-        action_log_probs = log_probs.gather(1, actions.view(-1, 1))
-        pi_err = -advantages.dot(action_log_probs.reshape(-1))
+    def forward(self, policy_model, observations, actions, advantages):
+        logits, _, values, _ = policy_model(
+            {SampleBatch.CUR_OBS: observations}, [])
+        dist = self.dist_class(logits)
+        log_probs = dist.logp(actions)
+        if len(log_probs.shape) > 1:
+            log_probs = log_probs.sum(-1)
```
Comment (ericl) [inline python/ray/rllib/agents/pg/torch_pg_policy_graph.py]: In which cases does log_probs have a nontrivial second dimension? Wondering if the reshape() is sufficient?

Same question for A3CLoss.

File: python/ray/rllib/evaluation/torch_policy_graph.py [hunk 5]
```diff
@@ -47,14 +48,20 @@ def __init__(self, observation_space, action_space, model, loss,
...
                 passed to the loss module's forward() function when computing
                 the loss. For example, ["obs", "action", "advantages"].
+            action_distribution_cls (ActionDistribution): Class for action
+                distribution.
         """
         self.observation_space = observation_space
         self.action_space = action_space
         self.lock = Lock()
-        self._model = model
+        cuda_devices = os.environ['CUDA_VISIBLE_DEVICES'].split(',')
```
Comment (ericl) [inline python/ray/rllib/evaluation/torch_policy_graph.py]: could simply check if `bool(os.environ.get("CUDA_VISIBLE_DEVICES"))`

File: python/ray/rllib/models/action_dist.py [hunk 6]
```diff
@@ -285,3 +286,40 @@ def kl(self, other):
     @override(ActionDistribution)
     def _build_sample_op(self):
         return self.dist.sample()
+
+
+class TorchDistributionWrapper(ActionDistribution):
```
Comment (ericl) [inline python/ray/rllib/models/action_dist.py]: Could the torch classes for action dist be placed in a separate file?

File: python/ray/rllib/models/action_dist.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 from collections import namedtuple
 import distutils.version
 import tensorflow as tf
+import torch
```
Comment (ericl) [inline python/ray/rllib/models/action_dist.py]: Let's make sure to not import torch unless we hit a torch=true code path, to avoid acquiring a hard dependency on torch.

File: python/ray/rllib/models/catalog.py [hunk 8]
```diff
@@ -120,7 +121,8 @@ def get_action_dist(action_space, config, dist_type=None):
             elif dist_type == "deterministic":
                 return Deterministic, action_space.shape[0]
         elif isinstance(action_space, gym.spaces.Discrete):
-            return Categorical, action_space.n
+            dist = TorchCategorical if torch else Categorical
```
Comment (ericl) [inline python/ray/rllib/models/catalog.py]: could we add `if torch: raise NotImplementedError` for the other dist types?

**Comment by ericl** [comment]
Thanks for opening this! Overall looks solid; have some comments.

File: python/ray/rllib/evaluation/torch_policy_graph.py [hunk 9]
```diff
@@ -67,45 +74,50 @@ def compute_actions(self,
...
         with self.lock:
             with torch.no_grad():
-                ob = torch.from_numpy(np.array(obs_batch)).float()
+                ob = torch.from_numpy(np.array(obs_batch)) \
+                    .float().to(self.device)
                 model_out = self._model({"obs": ob}, state_batches)
                 logits, _, vf, state = model_out
-                actions = F.softmax(logits, dim=1).multinomial(1).squeeze(0)
-                return (actions.numpy(), [h.numpy() for h in state],
+                action_dist = self._action_dist_cls(logits)
```
Comment (ericl) [inline python/ray/rllib/evaluation/torch_policy_graph.py]: Now that A2C/PG presumably work with continuous action spaces, you can add two entries to run_rllib_tests.sh to check they work on Pendulum-v0:

Similar to the CartPole-v0 entries:
https://github.com/ray-project/ray/blob/master/ci/jenkins_tests/run_rllib_tests.sh#L407

File: python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py (hunk 3 above)
Comment (FlyClover) [reply python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]: I can make an abstract method for getting extra grad info. 
For grad clipping, I can either make config a property of TorchPolicyGraph so compute_gradients() in TorchPolicyGraph would know whether to clip grad or make an abstract method extra_grad_processing(self, grad) in TorchPolicyGraph and let subclass process the grad. What's your preference?

File: python/ray/rllib/agents/pg/torch_pg_policy_graph.py (hunk 4 above)
Comment (FlyClover) [reply python/ray/rllib/agents/pg/torch_pg_policy_graph.py]: I haven't tried others but Normal distribution's log_prob returns vector of shape (n,) where n is the number of gaussians. I can absorb this into TorchDiagGaussian.

File: python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py]: TFPolicyGraph offers the extra grad processing method, so it's probably better to do that for consistency.

**Review by ericl — APPROVED** [review]
LGTM. One thing I'm wondering is if it's possible to test that GPU mode works properly, without a real GPU. It seems easy to forget a cpu().

**Comment by FlyClover** [comment]
Is it possible to spin up a gpu instance every night and run a nightly test build on it to catch some errors?

**Comment by ericl** [comment]
Hm, potentially. I'm not sure if travis supports GPU instances though.

@FlyClover tests look good, but you have a couple lint changes: https://travis-ci.com/ray-project/ray/jobs/192067752

**Comment by ericl** [comment]
Merged, thanks!
